cs-utilities/src/main/java/io/cloudslang/content/utilities/actions/OsDetector.java

```diff
@@ -0,0 +1,413 @@
+                    .withTrustAllRoots(defaultIfEmpty(trustAllRoots, valueOf(false)))
+                    .withX509HostnameVerifier(defaultIfEmpty(x509HostnameVerifier, X_509_HOSTNAME_VERIFIER.getValue()))
+                    .withTrustKeystore(trustKeystore)
+                    .withTrustPassword(trustPassword)
+                    .withKerberosConfFile(kerberosConfFile)
+                    .withKerberosLoginConfFile(kerberosLoginConfFile)
+                    .withKerberosSkipPortForLookup(kerberosSkipPortForLookup)
+                    .withKeystore(keystore)
+                    .withKeystorePassword(keystorePassword)
+                    .withWinrmLocale(defaultIfEmpty(winrmLocale, InputDefaults.WINRM_LOCALE.getValue()))
+                    .build();
+
+            OsDetectorUtils osDetectorUtils = new OsDetectorUtils();
+            NmapOsDetectorService nmapOsDetectorService = new NmapOsDetectorService(osDetectorUtils);
+            OperatingSystemDetector service = new OperatingSystemDetector(new SshOsDetectorService(osDetectorUtils, new ScoreSSHShellCommand()),
+                    new PowerShellOsDetectorService(osDetectorUtils, new WSManRemoteShellService()),
+                    nmapOsDetectorService,
+                    new LocalOsDetectorService(osDetectorUtils), osDetectorUtils);
+
+            validateNmapInputs(osDetectorInputs, nmapOsDetectorService);
+
+            OperatingSystemDetails os = service.detectOs(osDetectorInputs);
+            Map<String, String> returnResult;
+            if (osDetectorUtils.foundOperatingSystem(os)) {
+                returnResult = getSuccessResultsMap("Successfully detected the operating system.");
+                returnResult.put(OS_FAMILY, os.getFamily());
+                returnResult.put(OS_NAME, os.getName());
+                returnResult.put(OS_ARCHITECTURE, os.getArchitecture());
+                returnResult.put(OS_VERSION, os.getVersion());
+            } else {
+                returnResult = getFailureResultsMap("Unable to detect the operating system.");
+            }
+            returnResult.put(OS_COMMANDS, formatOsCommandsOutput(os.getCommandsOutput()));
+            return returnResult;
+        } catch (Exception e) {
+            return getFailureResultsMap(e);
+        }
+
+    }
+
+    private String formatOsCommandsOutput(Map<String, List<String>> commandsOutput) {
+        StringBuilder stringBuilder = new StringBuilder();
+        for (Map.Entry<String, List<String>> entry : commandsOutput.entrySet()) {
+            stringBuilder.append(entry.getKey()).append(DETECTION).append(lineSeparator());
+            for (String cmdOutput : entry.getValue()) {
+                stringBuilder.append(cmdOutput).append(lineSeparator());
+            }
+        }
+        return stringBuilder.toString();
+    }
+
+    private void validateNmapInputs(OsDetectorInputs osDetectorInputs, NmapOsDetectorService nmapOsDetectorService) {
```

Comment:
you should also move these two methods (`formatOsCommandsOutput`, `validateNmapInputs`) in another class, as it is a best practice to have only the `execute` method in the Action class